huaweicloud/services/modelarts/resource_huaweicloud_modelarts_workspace.go

```diff
@@ -0,0 +1,452 @@
+			},
+			"status_info": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: `Status details.`,
+			},
+		},
+	}
+}
+
+func modelartsWorkspaceGrantsSchema() *schema.Resource {
+	sc := schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"user_id": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: `IAM user ID.`,
+			},
+			"user_name": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: `IAM username.`,
+			},
+		},
+	}
+	return &sc
+}
+
+func resourceModelartsWorkspaceCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+
+	// createWorkspace: create a Modelarts workspace.
+	var (
+		createWorkspaceHttpUrl = "v1/{project_id}/workspaces"
+		createWorkspaceProduct = "modelarts"
+	)
+	createWorkspaceClient, err := cfg.NewServiceClient(createWorkspaceProduct, region)
+	if err != nil {
+		return diag.Errorf("error creating ModelArts Client: %s", err)
+	}
+
+	createWorkspacePath := createWorkspaceClient.Endpoint + createWorkspaceHttpUrl
+	createWorkspacePath = strings.ReplaceAll(createWorkspacePath, "{project_id}", createWorkspaceClient.ProjectID)
+
+	createWorkspaceOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+		OkCodes: []int{
+			200,
+		},
+	}
+
+	createWorkspaceOpt.JSONBody = utils.RemoveNil(buildCreateWorkspaceBodyParams(d, cfg))
+	createWorkspaceResp, err := createWorkspaceClient.Request("POST", createWorkspacePath, &createWorkspaceOpt)
+	if err != nil {
+		return diag.Errorf("error creating Modelarts workspace: %s", err)
+	}
+
+	createWorkspaceRespBody, err := utils.FlattenResponse(createWorkspaceResp)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	id, err := jmespath.Search("id", createWorkspaceRespBody)
+	if err != nil {
+		return diag.Errorf("error creating Modelarts workspace: ID is not found in API response")
+	}
+	d.SetId(id.(string))
+	return resourceModelartsWorkspaceRead(ctx, d, meta)
+}
+
+func buildCreateWorkspaceBodyParams(d *schema.ResourceData, cfg *config.Config) map[string]interface{} {
+	bodyParams := map[string]interface{}{
+		"name":                  utils.ValueIngoreEmpty(d.Get("name")),
+		"description":           utils.ValueIngoreEmpty(d.Get("description")),
+		"enterprise_project_id": utils.ValueIngoreEmpty(common.GetEnterpriseProjectID(d, cfg)),
+		"auth_type":             utils.ValueIngoreEmpty(d.Get("auth_type")),
+		"grants":                buildCreateWorkspaceRequestBodyGrants(d.Get("grants")),
+	}
+	return bodyParams
+}
+
+func buildCreateWorkspaceRequestBodyGrants(rawParams interface{}) []map[string]interface{} {
+	if rawArray, ok := rawParams.([]interface{}); ok {
+		if len(rawArray) == 0 {
+			return nil
+		}
+
+		rst := make([]map[string]interface{}, len(rawArray))
+		for i, v := range rawArray {
+			raw := v.(map[string]interface{})
```

Comment:
it's better to use "comma-ok" to avoid panic with ```grants {}```

Reply:
ok,done